File: 2/cpp/leetcode-2.cpp
#include <iostream>
#include <cmath>
class Solution {
public:
  int reverse(int x) {
    int ret = 0;
    
    while(x){
      int temp = ret * 10 + x % 10;
      if (ret != temp / 10) {
        //  已经溢出
        return 0;
      }
      ret = temp;
      x /= 10;
    }
    
    return ret;
  }
  int reverse2(int x) {
    int nums[100];
    int reverse[100];
    long ret = 0;
    int n = 0;
    
    while(x){
      /* code */
      nums[n++] = x % 10;
      x /= 10;
      std::cout << "x:::" << x << " nums:::" << nums[n-1] << " n:::" << n << std::endl;
    }
    int t = n;
    for (int i = 0; i < n; ++i) {
      ret += pow(10,--t)*nums[i];
      std::cout << "ret:::" << ret << " nums:::" << nums[i] << " t:::" << t << std::endl;
      if (ret > 0x7fffffff || ret < (signed int)0x80000000) {
        return 0;
      }
    }
    return ret;
  }
};

int main(void) {
  Solution s;
  int num = 1534236469;
  int ret = s.reverse2(num);
  std::cout << "before reverse::" << num << std::endl;
  std::cout << "reversed::" << ret << std::endl;
  return 0;
}

File: 2/javascript/leetcode-2.js
/**
 * @param {number} x
 * @return {number}
 */
function reverse(x) {
  let flag = x < 0 ? -1 : 1;
  x = x * flag;
  let str = String(x);
  let ret = Number(str.split("").reverse().join(""))*flag;
  if (ret > 2147483647 || ret < -2147483648) {
    return 0;
  }
  return ret;
}

function reverse2(x) {
  let ret = 0;
  while (x) {
    temp = ret * 10 + parseInt(x % 10);
    ret = temp;
    if (ret > 2147483647 || ret < -2147483648) {
      return 0;
    }
    x = parseInt(x/10);
  }
  return ret;
}

let ret = reverse2(-123);
console.log(ret);